skins/laika/src/components/pages/donation_confirmation/AddressModal.vue

```diff
@@ -0,0 +1,203 @@
+				lastName: {
+					name: 'lastName',
+					value: '',
+					pattern: '^.+$',
+					optionalField: false,
+				},
+				street: {
+					name: 'street',
+					value: '',
+					pattern: '^.+$',
+					optionalField: false,
+				},
+				city: {
+					name: 'city',
+					value: '',
+					pattern: '^.+$',
+					optionalField: false,
+				},
+				postcode: {
+					name: 'postcode',
+					value: '',
+					pattern: '^[0-9]{4,5}$',
+					optionalField: false,
+				},
+				country: {
+					name: 'country',
+					value: 'DE',
+					pattern: '',
+					optionalField: false,
+				},
+				email: {
+					name: 'email',
+					value: '',
+					pattern: '^(.+)@(.+)\\.(.+)$',
+					optionalField: false,
+				},
+			},
+		};
+	},
+	props: {
+		donation: Object,
+		updateDonorUrl: String,
+		validateAddressUrl: String,
+		countries: Array as () => Array<String>,
+		hasErrored: Boolean,
+		hasSucceeded: Boolean
+	},
+	computed: {
+		fieldErrors: {
+			get: function (): AddressValidity {
+				return Object.keys( this.formData ).reduce( ( validity: AddressValidity, fieldName: string ) => {
+					if ( !this.formData[ fieldName ].optionalField ) {
+						validity[ fieldName ] = this.$store.state.address.validity[ fieldName ] === Validity.INVALID;
+					}
+					return validity;
+				}, ( {} as AddressValidity ) );
+			},
+		},
+		...mapGetters( NS_ADDRESS, [
+			'addressType',
+		] ),
+	},
+	methods: {
+		validateForm(): Promise<ValidationResult> {
+			return this.$store.dispatch( action( NS_ADDRESS, validateAddress ), this.$props.validateAddressUrl );
+		},
+		onFieldChange( fieldName: string ): void {
+			this.$store.dispatch( action( NS_ADDRESS, setAddressField ), this.$data.formData[ fieldName ] );
+		},
+		setReceiptOptedOut( optedOut: boolean ): void {
+			this.$store.dispatch( action( NS_ADDRESS, setReceiptOptOut ), optedOut );
+		},
+		setAddressType( addressType: AddressTypeModel ): void {
+			this.$store.dispatch( action( NS_ADDRESS, setAddressType ), addressType );
+		},
+		submit() {
+			this.validateForm().then( ( validationResult: ValidationResult ) => {
+				if ( validationResult.status === 'OK' ) {
+					let form = this.$refs.form as HTMLFormElement;
+					trackFormSubmission( form );
+					const jsonForm = new FormData();
+					Object.keys( this.$data.formData ).forEach( fieldName => {
+						jsonForm.set( fieldName, this.$data.formData[ fieldName ].value );
+					} );
+					jsonForm.set( 'updateToken', this.$props.donation.updateToken );
+					jsonForm.set( 'donation_id', this.$props.donation.id );
+					jsonForm.set( 'addressType', addressTypeName( this.$store.getters[ NS_ADDRESS + '/addressType' ] ) );
+					axios.post(
+						this.$props.updateDonorUrl,
+						jsonForm,
+						{ headers: { 'Content-Type': 'multipart/form-data' } }
+					).then( ( validationResult: AxiosResponse<any> ) => {
+						if( validationResult.data.status === 'OK' ) {
+							this.$emit( 'address-updated' );
+						} else {
+							this.$emit( 'address-update-failed' );
+						}
```

Comment:
With the current event handler setup in `DonationConfirmation`, the modal will close when the server returns an error. This seems like a sensible behavior to me (because "retrying" it in the model probably won't help), but maybe we should pass the error message key in the event and display it in `DonationConfirmation`.

Reply:
I am not sure if you miswrote or if the behavior broke in the meantime but the modal will / should not close when an error is received. I think Jan mentioned this during a UX meeting and his reasoning was sort of like "the user should have time to read the text an move on on their own terms" so the modal stays up.

Reply:
I did not user-test that, I deducted the behavior from the code. Good to know, I'll user-test it then when the PR is finished.